jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionDataStore.java

```diff
@@ -76,47 +83,84 @@ public void initialize (SessionContext context) throws Exception
         _context = context;
     }
 
-    /** 
-     * @see org.eclipse.jetty.server.session.SessionDataStore#store(java.lang.String, org.eclipse.jetty.server.session.SessionData)
-     */
+    
+    
+    @Override
+    public SessionData load(String id) throws Exception
+    {
+        final AtomicReference<SessionData> reference = new AtomicReference<SessionData>();
+        final AtomicReference<Exception> exception = new AtomicReference<Exception>();
+        
+        Runnable r = new Runnable()
+        {
+            @Override
+            public void run ()
+            {
+                try
+                {
+                    reference.set(doLoad(id));
+                }
+                catch (Exception e)
+                {
+                    exception.set(e);
+                }
+            }
+        };
+
+        _context.run(r);
+        if (exception.get() != null)
+            throw exception.get();
+        
+        return reference.get();
+    }
```

Comment:
Or use a `FuturePromise`, which can do both the result or the exception.

Reply:
@gregw but it's not passed to another thread, so there is no "future" - it's just a method call with a return value and an exception.

Reply:
It if wrapped as a Runnable so it can't throw. We know it is run by the same thread but there is no guarantee of that. So there is a future.

Reply:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

Reply:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

Reply:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

Reply:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

Reply:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

Reply:
Sorry guys, nobody has made a convincing argument to change this code in any way: this method is in the base class for all *SessionDataStores. It is implemented in a parsimonious way with 2 AtomicReferences, one to capture the result and one to capture the exception: any of the other suggestions here result in a hell of lot more object creation for absolutely no gain whatsoever.

Reply:
@janbartel what I suggest will have no object creation at all!

Reply:
Well perhaps we can change `void ServletContext.run()` to `Object ServerContext.run() throws Exception`, but a) that is not really something for this PR; b) that will probably require new/changed catches elsewhere, which we need to review.    For now, this PR replaces several occurrences of this pattern with a single one, so even if it is ugly, it is better.